File: src/main/java/com/milaboratory/mixcr/cli/postanalysis/CommandPaExportPlots.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis;

import com.milaboratory.miplots.ExportKt;
import com.milaboratory.miplots.ExportType;
import com.milaboratory.mixcr.cli.CommonDescriptions;
import com.milaboratory.mixcr.postanalysis.plots.Filter;
import com.milaboratory.mixcr.postanalysis.plots.MetadataKt;
import com.milaboratory.util.StringUtil;
import jetbrains.letsPlot.intern.Plot;
import org.jetbrains.kotlinx.dataframe.DataFrame;
import org.jetbrains.kotlinx.dataframe.api.ToDataFrameKt;
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toList;

@Command(name = "exportPlots",
        separator = " ",
        description = "Export postanalysis plots.")
public abstract class CommandPaExportPlots extends CommandPaExport {
    @Option(description = CommonDescriptions.METADATA,
            names = {"--metadata"})
    public String metadata;

    @Option(description = "Plot width",
            names = {"--width"})
    public int width = 0;

    @Option(description = "Plot height",
            names = {"--height"})
    public int height = 0;

    @Option(description = "Filter by metadata. Possible filters column=value, column>=value etc.",
            names = {"--filter"},
            split = ",")
    public List<String> filterByMetadata;

    @Parameters(description = "Output PDF/EPS/PNG/JPEG file name", index = "1", defaultValue = "plot.pdf")
    public String out;

    @Override
    protected List<String> getOutputFiles() {
        return Collections.emptyList(); // output will be always overriden
    }

    protected <T> DataFrame<T> filter(DataFrame<T> df) {
        if (filterByMetadata != null)
            for (Filter f : filterByMetadata.stream().map(f -> MetadataKt.parseFilter(metadata(), f)).collect(Collectors.toList()))
                df = f.apply(df);
        return df;
    }

    private DataFrame<?> metadataDf;

    /** Get metadata from file */
    protected DataFrame<?> metadata() {
        if (metadataDf != null)
            return metadataDf;
        if (metadata != null)
            return metadataDf = MetadataKt.readMetadata(metadata);
        if (getPaResult().metadata != null)
            return metadataDf = ToDataFrameKt.toDataFrame(getPaResult().metadata);
        return null;
    }

    @Override
    public void validate() {
        super.validate();
        try {
            ExportType.determine(Paths.get(out));
        } catch (Exception e) {
            throwValidationException("Unsupported file extension (possible: pdf, eps, svg, png): " + out);
        }
        if (metadata != null && !metadata.endsWith(".csv") && !metadata.endsWith(".tsv"))
            throwValidationException("Metadata should be .csv or .tsv");
        if (metadata != null) {
            if (!metadata().containsColumn("sample"))
                throwValidationException("Metadata must contain 'sample' column");
            List<String> samples = getInputFiles();
            @SuppressWarnings("unchecked")
            Map<String, String> mapping = StringUtil.matchLists(
                    samples,
                    ((List<Object>) metadata().get("sample").toList())
                            .stream().map(Object::toString).collect(toList())
            );
            if (mapping.size() < samples.size() || mapping.values().stream().anyMatch(Objects::isNull))
                throwValidationException("Metadata samples does not match input file names.");
        }
        if (filterByMetadata != null && metadata() == null)
            throwValidationException("Filter is specified by metadata is not.");
    }

    String plotDestStr(IsolationGroup group) {
        String ext = out.substring(out.length() - 4);
        return out.substring(0, out.length() - 4) + group.extension() + ext;
    }

    Path plotDestPath(IsolationGroup group) {
        return Paths.get(plotDestStr(group));
    }

    protected void ensureOutputPathExists() {
        try {
            Files.createDirectories(Paths.get(out).toAbsolutePath().getParent());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

//    void writePlotsAndSummary(IsolationGroup group, List<byte[]> plots) {
//        ensureOutputPathExists();
//        ExportKt.writePDF(plotDestPath(group), plots);
//    }

    void writePlots(IsolationGroup group, List<Plot> plots) {
        ensureOutputPathExists();
        ExportKt.writeFile(plotDestPath(group), plots);
    }

    void writePlots(IsolationGroup group, Plot plot) {
        writePlots(group, Collections.singletonList(plot));
    }

    @Command(name = "exportPlots",
            separator = " ",
            description = "Export postanalysis plots.",
            subcommands = {
                    CommandLine.HelpCommand.class
            })
    public static class CommandExportPlotsMain {}
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/postanalysis/CommandPaExportPlots.kt
/*
 * Copyright (c) 2014-2024, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis

import com.milaboratory.app.InputFileType
import com.milaboratory.app.ValidationException
import com.milaboratory.miplots.writeFile
import com.milaboratory.mixcr.cli.CommonDescriptions
import com.milaboratory.mixcr.cli.CommonDescriptions.Labels
import com.milaboratory.mixcr.cli.exportTypes
import com.milaboratory.mixcr.postanalysis.plots.parseFilter
import com.milaboratory.mixcr.postanalysis.plots.readMetadata
import com.milaboratory.util.StringUtil
import jetbrains.letsPlot.intern.Plot
import org.jetbrains.kotlinx.dataframe.DataFrame
import org.jetbrains.kotlinx.dataframe.api.toDataFrame
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.io.path.absolute
import kotlin.io.path.extension
import kotlin.io.path.nameWithoutExtension

abstract class CommandPaExportPlots : CommandPaExport() {
    @set:Option(
        description = [CommonDescriptions.METADATA],
        names = ["--metadata"],
        paramLabel = "<path.tsv>",
        order = OptionsOrder.main + 4_000
    )
    var metadata: Path? = null
        set(value) {
            ValidationException.requireFileType(value, InputFileType.TSV)
            ValidationException.requireFileExists(value)
            field = value
        }

    @Option(
        description = ["Plot width."],
        names = ["--width"],
        paramLabel = "<n>",
        order = OptionsOrder.width
    )
    var width = 0

    @Option(
        description = ["Plot height."],
        names = ["--height"],
        paramLabel = "<n>",
        order = OptionsOrder.height
    )
    var height = 0

    @Option(
        description = ["Filter samples to put on a plot by their metadata values. Filter allows equality (`species=cat`) or arithmetic comparison (`age>=10`) etc."],
        names = ["--filter"],
        split = ",",
        paramLabel = "<meta(|>|>=|=|<=|<)value>",
        order = OptionsOrder.main + 4_100
    )
    var filterByMetadata: List<String>? = null

    @Parameters(
        description = ["Output PDF/EPS/PNG/JPEG file name."],
        index = "1",
        paramLabel = "output.${Labels.EXPORT_TYPES}"
    )
    lateinit var out: Path

    protected fun <T> DataFrame<T>.filterByMetadata(): DataFrame<T> {
        var result = this
        filterByMetadata?.let { filterByMetadata ->
            for (f in filterByMetadata.map { f -> metadataDf!!.parseFilter(f) }) {
                result = f.apply(result)
            }
        }
        return result
    }

    /** Get metadata from file  */
    protected val metadataDf: DataFrame<*>? by lazy {
        when {
            metadata != null -> readMetadata(metadata!!)
            else -> paResult.metadata?.toDataFrame()
        }
    }

    override fun validate() {
        super.validate()
        ValidationException.requireFileType(out, InputFileType.exportTypes)
        metadataDf?.let { metadataDf ->
            if (!metadataDf.containsColumn("sample"))
                throw ValidationException("Metadata must contain 'sample' column")
            val samples = inputFiles
            val mapping = StringUtil.matchLists(
                samples.map { it.toString() },
                metadataDf["sample"].toList().map { it!!.toString() }
            )
            if (mapping.size < samples.size || mapping.values.any { it == null })
                throw ValidationException("Metadata samples does not match input file names.")
        }
        if (filterByMetadata != null && metadataDf == null)
            throw ValidationException("Filter is specified by metadata is not.")
    }

    private fun plotDestStr(group: IsolationGroup): String {
        val out = out.absolute()
        val ext = out.extension
        val withoutExt = out.parent.resolve(out.nameWithoutExtension)
        return "$withoutExt${group.extension()}.$ext"
    }

    private fun plotDestPath(group: IsolationGroup): Path = Paths.get(plotDestStr(group))

    private fun ensureOutputPathExists() {
        Files.createDirectories(out.toAbsolutePath().parent)
    }

    //    void writePlotsAndSummary(IsolationGroup group, List<byte[]> plots) {
    //        ensureOutputPathExists();
    //        ExportKt.writePDF(plotDestPath(group), plots);
    //    }
    fun writePlots(group: IsolationGroup, plots: List<Plot>) {
        ensureOutputPathExists()
        writeFile(plotDestPath(group), plots)
    }

    fun writePlots(group: IsolationGroup?, plot: Plot) {
        writePlots(group!!, listOf(plot))
    }

    @Command(
        description = ["Export postanalysis plots."],
        synopsisSubcommandLabel = "COMMAND"
    )
    class CommandExportPlotsMain
}
